appsec/dyngo/events_test.go

```diff
@@ -0,0 +1,238 @@
+package dyngo_test
+
+import (
+	"fmt"
+	"github.com/DataDog/dd-trace-go/appsec/dyngo"
+	"reflect"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+type (
+	testOp1Args struct{}
+	testOp1Res  struct{}
+
+	testOp2Args struct{}
+	testOp2Res  struct{}
+
+	testOp3Args struct{}
+	testOp3Res  struct{}
+)
+
+func TestOperationEvents(t *testing.T) {
+	t.Run("single operation without listeners", func(t *testing.T) {
+		require.NotPanics(t, func() {
+			op := dyngo.StartOperation(testOp1Args{})
+			op.Finish(testOp1Res{})
+		})
+
+		require.Panics(t, func() {
+			op := dyngo.StartOperation(nil)
+			op.Finish(nil)
+		})
+	})
+
+	t.Run("start event", func(t *testing.T) {
+		op1 := dyngo.StartOperation(testOp1Args{})
+
+		var called int
+		op1.OnStart(func(op *dyngo.Operation, args testOp2Args) {
+			called++
+		})
+
+		// Not called
+		require.Equal(t, 0, called)
+
+		op2 := dyngo.StartOperation(testOp2Args{}, dyngo.WithParent(op1))
+		op2.Finish(testOp2Res{})
+
+		// Called once
+		require.Equal(t, 1, called)
+
+		op2 = dyngo.StartOperation(testOp2Args{}, dyngo.WithParent(op1))
+		op2.Finish(testOp2Res{})
+
+		// Called again
+		require.Equal(t, 2, called)
+
+		// Finish the operation so that it gets disabled and its listeners removed
+		op1.Finish(testOp1Res{})
+
+		op2 = dyngo.StartOperation(testOp2Args{}, dyngo.WithParent(op1))
+		op2.Finish(testOp2Res{})
+
+		// No longer called
+		require.Equal(t, 2, called)
+	})
+
+	t.Run("finish event", func(t *testing.T) {
+		op1 := dyngo.StartOperation(testOp1Args{})
+
+		var called int
+		op1.OnFinish(func(op *dyngo.Operation, args testOp2Res) {
+			called++
+		})
+
+		op2 := dyngo.StartOperation(testOp2Args{}, dyngo.WithParent(op1))
+		op2.Finish(testOp2Res{})
+		// Called once
+		require.Equal(t, 1, called)
+
+		op2 = dyngo.StartOperation(testOp2Args{}, dyngo.WithParent(op1))
+		op2.Finish(testOp2Res{})
+		// Called again
+		require.Equal(t, 2, called)
+
+		op3 := dyngo.StartOperation(testOp3Args{}, dyngo.WithParent(op2))
+		op3.Finish(testOp3Res{})
+		// Not called
+		require.Equal(t, 2, called)
+
+		op2 = dyngo.StartOperation(testOp2Args{}, dyngo.WithParent(op3))
+		op2.Finish(testOp2Res{})
+		// Called again
+		require.Equal(t, 3, called)
+
+		// Finish the operation so that it gets disabled and its listeners removed
+		op1.Finish(testOp1Res{})
+
+		op2 = dyngo.StartOperation(testOp2Args{}, dyngo.WithParent(op1))
+		op2.Finish(testOp2Res{})
+		// No longer called
+		require.Equal(t, 3, called)
+	})
+}
+
+func BenchmarkGoAssumptions(b *testing.B) {
+	// Compare map lookup times according to their key type.
+	// The selected implementation assumes using reflect.TypeOf(v).Name() doesn't allocate memory
+	// and is as good as "regular" string keys, whereas the use of reflect.Type keys is slower due
+	// to the underlying struct copy of the reflect struct type descriptor which has a lot of
+	// fields copied involved in the key comparison.
+	b.Run("map lookups", func(b *testing.B) {
```

Comment:
@felixge: here are some benchmarks of map lookup times according to their key type.
In the end, reflect.Type is faster than the custom type struct. But the difference is a few nanoseconds that are imo completely negligible.